File: www/hubic.js
var url = require("url");
var path = require("path");

var request = require("./request");
var oauth = require("./oauth");

var CREDENTIALS_ENDPOINT = "https://api.hubic.com/1.0/account/credentials";

function token(session, callback, code, refresh) {
  if (typeof code !== "string" || !code) {
    return callback ? callback() : undefined;
  }

  oauth.token(session, oauth.SERVICE.HUBIC, code, refresh, callback);
}

function credentials(session, callback, access_token) {
  if (typeof access_token !== "string") {
    return callback ? callback() : undefined;
  }

  var headers = {
    "Authorization": "Bearer " + access_token
  };

  var u = url.parse(CREDENTIALS_ENDPOINT);

  request("GET", u.hostname, u.pathname, headers, null, null, callback);
}

function getObject(session, callback, token, endpoint, resource_path) {
  if (typeof token !== "string" || typeof endpoint !== "string" ||
    typeof resource_path !== "string") {
    return callback ? callback() : undefined;
  }

  while (resource_path.charAt(0) === "/") {
    resource_path = resource_path.slice(1);
  }

  var headers = {
    "X-Auth-Token": token
  };

  var u = url.parse(endpoint);

  if (typeof u.hostname !== "string" || typeof u.pathname !== "string" ||
    u.hostname.slice(-1 * "hubic.ovh.net".length) !== "hubic.ovh.net") {
    return callback ? callback() : undefined;
  }

  request("GET", u.hostname, path.join(u.pathname, resource_path) +
    "?format=json", headers, null,
    null, callback);
}

function putObject(session, callback, token, endpoint, resource_path, content) {
  if (typeof token !== "string" || typeof endpoint !== "string" ||
    typeof resource_path !== "string" || typeof content !== "string") {
    return callback ? callback() : undefined;
  }

  while (resource_path.charAt(0) === "/") {
    resource_path = resource_path.slice(1);
  }

  var headers = {
    "X-Auth-Token": token,
    "Content-Type": "text/plain; charset=utf-8"
  };

  var u = url.parse(endpoint);

  if (typeof u.hostname !== "string" || typeof u.pathname !== "string" ||
    u.hostname.slice(-1 * "hubic.ovh.net".length) !== "hubic.ovh.net") {
    return callback ? callback() : undefined;
  }

  request("PUT", u.hostname, path.join(u.pathname, resource_path), headers,
    null, new Buffer(
      content, "utf8"), callback);
}

module.exports = {
  token: token,
  credentials: credentials,
  getObject: getObject,
  putObject: putObject
};

File: android/src/app/src/main/java/com/example/agitodo/Hubic.java
package com.example.agitodo;

import java.util.ArrayList;
import java.util.List;

import org.apache.http.message.BasicNameValuePair;
import org.apache.http.NameValuePair;

import android.net.Uri;

public class Hubic {
    static final String CREDENTIALS_ENDPOINT = "https://api.hubic.com/1.0/account/credentials";

    static Logger log = new Logger();

    String token;

    public Hubic(String token) {
        this.token = token;
    }

    public Request.Response credentials() {
        List<NameValuePair> headers = new ArrayList<NameValuePair>();
        headers.add(new BasicNameValuePair("Authorization", "Bearer " + token));

        return new Request(Request.GET, Uri.parse(CREDENTIALS_ENDPOINT), headers)
                .send(null);
    }

    public Request.Response getObject(String endpoint, String resource_path) {
        while (endpoint.endsWith("/")) {
            endpoint = endpoint.substring(0, endpoint.length() - 1);
        }

        while (resource_path.startsWith("/")) {
            resource_path = resource_path.substring(1);
        }

        List<NameValuePair> headers = new ArrayList<NameValuePair>();
        headers.add(new BasicNameValuePair("X-Auth-Token", token));

        Uri.Builder uriBuilder = Uri.parse(endpoint + "/" + resource_path)
                .buildUpon();
        uriBuilder.appendQueryParameter("format", "json");

        return new Request(Request.GET, uriBuilder.build(), headers).send(null);
    }

    public Request.Response putObject(String endpoint, String resource_path,
                                      String content) {
        while (endpoint.endsWith("/")) {
            endpoint = endpoint.substring(0, endpoint.length() - 1);
        }

        while (resource_path.startsWith("/")) {
            resource_path = resource_path.substring(1);
        }

        List<NameValuePair> headers = new ArrayList<NameValuePair>();
        headers.add(new BasicNameValuePair("X-Auth-Token", token));
        headers.add(new BasicNameValuePair("Content-Type",
                "text/plain; charset=utf-8"));

        return new Request(Request.PUT, Uri.parse(endpoint + "/" + resource_path),
                headers).send(content);
    }
}
